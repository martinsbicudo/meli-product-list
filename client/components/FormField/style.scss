.FormField {
  flex: 1;
  background-color: map-get($colors, "meli-white");
  border-radius: .2rem;
  border: .1rem solid transparent;
  outline: 0;

  &Input {
    background: inherit;
    width: inherit;
    border: 0;
    padding: 1rem;
    font-size: 1.5rem;
    box-sizing: border-box;
    width: 100%;
  }

  &Focus { border-color: map-get($colors, "meli-denim"); }
}
