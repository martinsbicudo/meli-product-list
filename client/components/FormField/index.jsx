import React, { useState } from 'react';
import { object, func } from 'prop-types';
import classnames from 'classnames';
import style from './style.scss';

const FormField = ({
  className,
  onFocus,
  onBlur,
  ...props
}) => {
  const [isFocus, setIsFocus] = useState(false);

  function toggleFocus(e) {
    if (isFocus && onBlur) onBlur(e);
    if (!isFocus && onFocus) onFocus(e);

    setIsFocus((currentIsFocus) => !currentIsFocus);
  }

  return (
    <div
      className={classnames(
        style.FormField,
        className.wrapper || '',
        { [style.FormFieldFocus]: isFocus },
      )}
    >
      <input
        className={classnames(
          style.FormFieldInput,
          className.input || '',
        )}
        onFocus={toggleFocus}
        onBlur={toggleFocus}
        {...props}
        type="text"
      />
    </div>
  );
};

FormField.propTypes = {
  className: object,
  onFocus: func,
  onBlur: func,
};

FormField.defaultProps = {
  className: {
    wrapper: '',
    input: '',
  },
  onFocus: undefined,
  onBlur: undefined,
};

export default FormField;
