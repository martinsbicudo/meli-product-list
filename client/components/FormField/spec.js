import FormField from '.';

describe('Components - <FormField />', () => {
  const testMessagePlaceholder = 'TEST_PLACEHOLDER';
  const spyOnFocus = jest.fn();
  const spyOnBlur = jest.fn();
  const spyOnChange = jest.fn();
  const initialState = {
    onFocus: spyOnFocus,
    onBlur: spyOnBlur,
    onChange: spyOnChange,
    placeholder: testMessagePlaceholder,
  };
  const createWrapper = (
    props = {},
  ) => render(
    <FormField
      {...{
        ...initialState,
        ...props,
      }}
    />,
  );

  test('Should calls onFocus', async () => {
    const { getByPlaceholderText } = createWrapper();

    fireEvent.focus(getByPlaceholderText(testMessagePlaceholder));

    expect(spyOnFocus).toHaveBeenCalledTimes(1);
    spyOnFocus.mockRestore();
  });

  test('Should calls onBlur', async () => {
    const { getByPlaceholderText } = createWrapper({
      onChange: () => {},
    });

    fireEvent.focus(getByPlaceholderText(testMessagePlaceholder));
    fireEvent.blur(getByPlaceholderText(testMessagePlaceholder));

    expect(spyOnBlur).toHaveBeenCalledTimes(1);
    spyOnBlur.mockRestore();
  });

  test('Should calls onChange', async () => {
    const { getByPlaceholderText } = createWrapper();

    fireEvent.change(getByPlaceholderText(testMessagePlaceholder), {
      target: {
        value: 'TEST_VALUE',
      },
    });

    expect(spyOnChange).toHaveBeenCalledTimes(1);
    spyOnChange.mockRestore();
  });

  test('Wrapper Snapshot', () => {
    const { asFragment } = createWrapper();

    expect(asFragment()).toMatchSnapshot();
  });
});
