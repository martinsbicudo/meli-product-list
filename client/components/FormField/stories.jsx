import React from 'react';
import { storiesOf } from '@storybook/react';
import FormField from '.';

storiesOf('FormField', module)
  .add('With onFocus', () => (
    <FormField
      onFocus={() => console.log('onFocus was called') /* eslint-disable-line */}
    />
  ))
  .add('With onBlur', () => (
    <FormField
      onBlur={() => console.log('onBlur was called') /* eslint-disable-line */}
    />
  ))
  .add('With onChange', () => (
    <FormField
      onChange={() => console.log('onChange was called') /* eslint-disable-line */}
    />
  ))
  .add('With Placeholder', () => (
    <FormField placeholder="Placeholder FormField" />
  ));
