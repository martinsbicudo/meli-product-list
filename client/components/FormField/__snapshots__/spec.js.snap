// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - <FormField /> Wrapper Snapshot 1`] = `
<DocumentFragment>
  <div
    class="FormField"
  >
    <input
      class="FormFieldInput"
      placeholder="TEST_PLACEHOLDER"
      type="text"
    />
  </div>
</DocumentFragment>
`;
